hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/TransactionManager.java

```diff
@@ -55,28 +55,32 @@ public void beginTransaction() {
 
   public void beginTransaction(Option<HoodieInstant> currentTxnOwnerInstant,
                                Option<HoodieInstant> lastCompletedTxnOwnerInstant) {
-    if (supportsOptimisticConcurrency) {
+    if (isOptimisticConcurrencyControlEnabled) {
       LOG.info("Transaction starting for " + currentTxnOwnerInstant
           + "with latest completed transaction instant " + lastCompletedTxnOwnerInstant);
       lockManager.lock();
-      this.currentTxnOwnerInstant = currentTxnOwnerInstant;
-      this.lastCompletedTxnOwnerInstant = lastCompletedTxnOwnerInstant;
+      reset(currentTxnOwnerInstant, lastCompletedTxnOwnerInstant);
       LOG.info("Transaction started for " + currentTxnOwnerInstant
           + "with latest completed transaction instant " + lastCompletedTxnOwnerInstant);
     }
   }
 
   public void endTransaction() {
-    if (supportsOptimisticConcurrency) {
+    if (isOptimisticConcurrencyControlEnabled) {
```

Comment:
help me understand something. lets say writer1 acquires the lock and takes lot of time to release. writer2 tries to acquire the lock, but times out. In finally block of any transaction handling code, we do end transaction right. In this case when writer2 fails to acquire, will end transaction be called? 
if yes, wouldn't writer2 resets the transaction owner at L 71

Reply:
writer2 end transaction will fail as he doesn't hold the lock

Reply:
yes, but the failure happens only at L 72 right.

Reply:
Good catch, will close the gap in the reset with CAS like operation.

Reply:
Addressed your comments in the latest patch revision. Added a test. Please take a look.